### PR TITLE
DDP + C10D sparse all_reduce changes

### PR DESCRIPTION
**Pull request by H-Huang**
Summary:
## Changes

prototyping sparse allreduce using the sparse dispatch key. When passing in sparse tensors into `dist.allreduce()` we can execute our dispatched function.

prior to this change, passing a sparse tensor into `allreduce()` will error out with `Tensor must be dense...`

## Example script

```python
# python -m torch.distributed.run --nnodes=1 --nproc_per_node=2 this_script.py

import torch
import torch.distributed as dist

def main():
    dist.init_process_group(backend="nccl")
    rank = dist.get_rank()
    a = torch.tensor([[0, 2.], [3, 0]]).to(rank)
    a = a.to_sparse()
    print(f"rank {rank} - a: {a}")
    dist.all_reduce(a)

if __name__ == "__main__":
    main()
```

output:
```
rank 1 - a: tensor(indices=tensor([[0, 1],
                       [1, 0]]),
       values=tensor([2., 3.]),
       device='cuda:1', size=(2, 2), nnz=2, layout=torch.sparse_coo)
allreduce_sparse_cuda_
tensor.is_sparse() = 1
in ProcessGroupNCCL::allreduceSparse
rank 0 - a: tensor(indices=tensor([[0, 1],
                       [1, 0]]),
       values=tensor([2., 3.]),
       device='cuda:0', size=(2, 2), nnz=2, layout=torch.sparse_coo)
allreduce_sparse_cuda_
tensor.is_sparse() = 1
in ProcessGroupNCCL::allreduceSparse
```

Test Plan:
Testing commands (OSS):

```
# python
pytest test/distributed/test_c10d_nccl.py -vsk test_sparse_allreduce_ops

# c++
build/bin/ProcessGroupNCCLTest --gtest_filter=ProcessGroupNCCLTest.testSparseAllreduce
```

Testing commands (internal, ondemand GPU):
ddp tests:
```
buck build mode/opt -c hpc_comms.use_nccl=exp //caffe2/test/distributed:c10d --show-full-output

# Get the .par file from the previous command and use it below
TORCH_SHOW_CPP_STACKTRACE=1 /data/sandcastle/boxes/fbsource/buck-out/v2/gen/fbcode/c8344b52091f4f7f/caffe2/test/distributed/__c10d__/c10d.par -r test_ddp_set_sparse_metadata
```

c10d tests:
```
# build tests and run with log output (python)
buck build mode/opt -c hpc_comms.use_nccl=exp //caffe2/test/distributed:c10d --show-full-output
NCCL_DEBUG=WARN /data/sandcastle/boxes/fbsource/buck-out/v2/gen/fbcode/c8344b52091f4f7f/caffe2/test/distributed/__c10d__/c10d.par -r test_sparse_allreduce_ops

# python
NCCL_DEBUG=WARN buck test mode/opt -c hpc_comms.use_nccl=exp //caffe2/test/distributed:c10d -- --exact 'caffe2/test/distributed:c10d - test_sparse_allreduce_ops (test_c10d_nccl.ProcessGroupNCCLTest)'

# c++
NCCL_DEBUG=WARN buck run mode/opt -c hpc_comms.use_nccl=exp //caffe2/test/cpp/c10d:ProcessGroupNCCLTest -- --gtest_filter=ProcessGroupNCCLTest.testSparseAllreduce
```

Differential Revision: D46724856

Pulled By: H-Huang

### CODE REVIEW

File: torch/nn/parallel/distributed.py
Comment (H-Huang) [inline torch/nn/parallel/distributed.py:1238]: @rohan-varma, is it okay to remove this check? Param to name mapping is needed for sparse allreduce when global unique id is passed in. I assume cost to create this mapping is negligible?

File: test/distributed/test_c10d_nccl.py
Comment (rohan-varma) [inline test/distributed/test_c10d_nccl.py:320]: do we want to try with `get_future()` API?

File: test/cpp/c10d/ProcessGroupNCCLTest.cpp
Comment (rohan-varma) [inline test/cpp/c10d/ProcessGroupNCCLTest.cpp:890]: does the test work?

File: test/distributed/test_c10d_nccl.py
Comment (rohan-varma) [inline test/distributed/test_c10d_nccl.py:3104]: nit: remove comments before landing

File: test/distributed/test_c10d_nccl.py
Comment (rohan-varma) [inline test/distributed/test_c10d_nccl.py:3136]: can we validate equivalence somehow? i.e. gradients are the same with and without sparse allreduce / setting sparse metadata?

File: torch/csrc/distributed/c10d/NCCLUtils.hpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/NCCLUtils.hpp:17]: where is it used?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1560]: don't we need this fixed, since we need to support sparse tensors here?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1569]: nit: I think the lack of this was a bug that stumped us for a while, so can you add a comment why this is needed? 

IIRC - since the tensor will be used on a different (the nccl) stream, we need to call recordStream to ensure the tensor is not prematurely freed?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1830]: what is the `note:` for?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1862]: nit: not needed to change in this PR, but theoretically we can just record `output` for every `collective` in the `collective` implementation?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1893]: remove?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1900]: maybe make this `nccl:all_reduce_sparse` for the profiler to distinguish

File: torch/csrc/distributed/c10d/comm.hpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/comm.hpp:28]: Will defer to you, but `global_unique_id` seems like a higher level implementation detail that doesn't fully reflect how users can use this variable. Can we name it something more conceptual / easier to understand?

File: torch/csrc/distributed/c10d/comm.hpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/comm.hpp:75]: pls check if it needs a lock

File: torch/csrc/distributed/c10d/default_comm_hooks.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/default_comm_hooks.cpp:53]: i might be confused, but does this mean we always call sparse allreduce if IS_NCCL_EXP = true (even if we have a dense tensor)? 

Have we tested the case where IS_NCCL_EXP=true and we're allreducing a dense tensor?

also I'm confused on this - if we have the dispatcher, why do we need to call `allreduce_sparse` explicitly here, shouldn't we call just `allreduce` and the dispatcher figures out which implementation to call?

File: torch/csrc/distributed/c10d/reducer.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.cpp:444]: since we're turning this on by default now, the check can probably be moved to the Reducer constructor. Here, it would probably be called on every parameter every iteration.

File: torch/csrc/distributed/c10d/reducer.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.cpp:448]: nit: we also access the map on L452, maybe read it only once and save it

File: torch/csrc/distributed/c10d/reducer.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.cpp:484]: what's this for?

File: torch/csrc/distributed/c10d/reducer.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.cpp:493]: what argument is this? can we just default it?

File: torch/csrc/distributed/c10d/reducer.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.cpp:1563]: is it possible for sparse_metadata = true and bucket.global_unique_id.has_value = False? In what cases can that happen?

File: torch/nn/parallel/distributed.py
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:1238]: yep

File: torch/csrc/distributed/c10d/reducer.hpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.hpp:508]: nit: I would add a description of what the key here is and what the value is conceptually (i.e. I feel like `global_unique_id` is an unfamiliar term to most users). The API is private for now but might be good for conceptual undestanding.

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks for adding this! Have a few comments inline. Main asks: 

1. Can we add more comprehensive unittesting - such as DDP + set_sparse_metadata + sparse allreduce, and verify (up to a tolerenance) with DDP + the same tensor, but dense? 
2. What happens now if I do DDP with sparse grads + PG NCCL, but don't call set_sparse_metadata? Are we supporting this feature, and if so are we sure about the correctness? If we're not sure about correctness we should probably just error out for now, if the scope of the PR is to just support with set_sparse_metadata. If adding tests for this is not in scope for this PR, let's file a follow up issue 
3. Consider renaming `global_unique_id` to something more conceptual and adding a docstr on it
4. I don't completely understand why we need to change the allreduce in default_comm_hooks.cpp and case it on USE_NCCL_EXP. Could you clarify what happens if we end up in that codepath, USE_NCCL_EXP=true, but we're allreducing a dense tensor? And why we need to distinguish there btwn allreduce_sparse and allreduce_, instead of using the dispatcher.

File: test/cpp/c10d/ProcessGroupNCCLTest.cpp
Comment (H-Huang) [reply test/cpp/c10d/ProcessGroupNCCLTest.cpp:890]: yes, both work and operate on different sizes. Will uncomment it

File: torch/csrc/distributed/c10d/NCCLUtils.hpp
Comment (H-Huang) [reply torch/csrc/distributed/c10d/NCCLUtils.hpp:17]: not used anymore, cleaning it up, thanks!

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp
Comment (H-Huang) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1560]: old TODO that isn't relevant anymore

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp
Comment (H-Huang) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1900]: nice catch, thanks!

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp
Comment (H-Huang) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1830]: thanks, not needed

File: torch/csrc/distributed/c10d/reducer.cpp
Comment (H-Huang) [reply torch/csrc/distributed/c10d/reducer.cpp:484]: updated so it isnt needed anymore

File: torch/csrc/distributed/c10d/default_comm_hooks.cpp
Comment (H-Huang) [reply torch/csrc/distributed/c10d/default_comm_hooks.cpp:53]: This is a good catch. I'm updating the code to only call into allreduce and utilize the dispatcher

File: torch/csrc/distributed/c10d/reducer.cpp
Comment (H-Huang) [reply torch/csrc/distributed/c10d/reducer.cpp:1563]: there shouldn't be a case, so yeah the check seems redundant

**Pull request closed without merging**
